### PR TITLE
Crypto - server side encryption

### PR DESCRIPTION
**Pull request by chenshay**

### CODE REVIEW

File: src/api/subscriptions.js [hunk 1]
```diff
@@ -59,9 +59,10 @@ export class Subscriptions {
   clear() {}
 
   /**
+   * @param {?string} opt_encryptedDocumentKey
```
Comment (dvoytenko) [inline src/api/subscriptions.js]: `?string=` for `opt_` parameters

File: src/runtime/entitlements-manager.js [hunk 2]
```diff
@@ -129,11 +129,13 @@ export class EntitlementsManager {
   }
 
   /**
+   * @param {?string} opt_encryptedDocumentKey
```
Comment (dvoytenko) [inline src/runtime/entitlements-manager.js]: ditto

File: src/runtime/entitlements-manager.js [hunk 3]
```diff
@@ -150,11 +152,13 @@ export class EntitlementsManager {
   /**
    * @param {string} raw
    * @param {boolean=} opt_isReadyToPay
+   * @param {?string} opt_decryptedDocumentKey
    * @return {boolean}
    */
-  pushNextEntitlements(raw, opt_isReadyToPay) {
+  pushNextEntitlements(raw, opt_isReadyToPay, opt_decryptedDocumentKey) {
```
Comment (dvoytenko) [inline src/runtime/entitlements-manager.js]: I don't think a decrypted document key could ever be available for the `push` operation. This comes from the payments flow which is completely document-agnostic.

File: src/runtime/entitlements-manager.js [hunk 4]
```diff
@@ -163,29 +167,33 @@ export class EntitlementsManager {
...
+  fetchEntitlementsWithCaching_(opt_encryptedDocumentKey) {
     return Promise.all([
       this.storage_.get(ENTS_STORAGE_KEY),
       this.storage_.get(IS_READY_TO_PAY_STORAGE_KEY),
     ]).then(cachedValues => {
       const raw = cachedValues[0];
       const irtp = cachedValues[1];
       // Try cache first.
-      if (raw) {
+      //todo(chenshay)
```
Comment (dvoytenko) [inline src/runtime/entitlements-manager.js]: ?

File: src/runtime/entitlements-manager.js [hunk 5]
```diff
@@ -310,17 +322,19 @@ export class EntitlementsManager {
    * @param {string} raw
    * @param {!Object|!Array<!Object>} json
    * @param {boolean=} opt_isReadyToPay
+   * @param {?string} opt_decryptedDocumentKey
```
Comment (dvoytenko) [inline src/runtime/entitlements-manager.js]: ditto

File: src/runtime/entitlements-manager.js [hunk 6]
```diff
@@ -390,15 +404,18 @@ export class EntitlementsManager {
   }
 
   /**
+   * @param {?string} opt_encryptedDocumentKey
```
Comment (dvoytenko) [inline src/runtime/entitlements-manager.js]: ditto

File: src/runtime/entitlements-manager.js [hunk 7]
```diff
@@ -390,15 +404,18 @@ export class EntitlementsManager {
...
    */
-  fetch_() {
-    const url = serviceUrl(
-        '/publication/' +
-        encodeURIComponent(this.publicationId_) +
-        '/entitlements');
-    return this.fetcher_.fetchCredentialedJson(url)
+  fetch_(opt_encryptedDocumentKey) {
+    let url = '/publication/' +
+    encodeURIComponent(this.publicationId_) +
```
Comment (dvoytenko) [inline src/runtime/entitlements-manager.js]: some indent issues here.

File: src/runtime/entitlements-manager.js [hunk 8]
```diff
@@ -390,15 +404,18 @@ export class EntitlementsManager {
...
+    '/entitlements';
+    if (opt_encryptedDocumentKey) {
+      url = url + '/crypt/' + opt_encryptedDocumentKey;
```
Comment (dvoytenko) [inline src/runtime/entitlements-manager.js]: Hmm. I don't see why it should be in the path - this is a non-identifying parameter. It should be a query param (and later maybe even a POST param).

Another critical issue: this parameter must be encoded for URL.

File: src/api/subscriptions.js (hunk 1 above)
Comment (chenshay) [reply src/api/subscriptions.js]: done

File: src/runtime/entitlements-manager.js (hunk 2 above)
Comment (chenshay) [reply src/runtime/entitlements-manager.js]: done

File: src/runtime/entitlements-manager.js (hunk 3 above)
Comment (chenshay) [reply src/runtime/entitlements-manager.js]: done

File: src/runtime/entitlements-manager.js (hunk 5 above)
Comment (chenshay) [reply src/runtime/entitlements-manager.js]: done

File: src/runtime/entitlements-manager.js (hunk 6 above)
Comment (chenshay) [reply src/runtime/entitlements-manager.js]: done

File: src/runtime/entitlements-manager.js (hunk 7 above)
Comment (chenshay) [reply src/runtime/entitlements-manager.js]: done

**Pull request merged**